rustls/src/conn.rs

```diff
@@ -862,57 +878,197 @@ impl<Data> ConnectionCore<Data> {
     fn deframe<'b>(
         &mut self,
         state: Option<&dyn State<Data>>,
-        deframer_buffer: &mut DeframerSliceBuffer<'b>,
+        buffer: &'b mut [u8],
+        buffer_range: &mut BufferRange,
+    ) -> Result<Option<InboundPlainMessage<'b>>, Error> {
+        // before processing any more of `buffer`, return any extant messages from `hs_deframer`
+        if let HandshakeDeframerStatus::Ready(_) = self.hs_deframer.status() {
+            Ok(Some(self.take_handshake_message(buffer, buffer_range)))
+        } else {
+            self.process_more_input(state, buffer, buffer_range)
+        }
+    }
+
+    fn take_handshake_message<'b>(
+        &mut self,
+        buffer: &'b mut [u8],
+        buffer_range: &mut BufferRange,
+    ) -> InboundPlainMessage<'b> {
+        // unwrap: guaranteed by caller checking `hs_deframer.status()`
+        let (message, discard) = self
```

Comment:
This fn body feels small enough that it might be safer to in-line it at the call-site so that the pre-condition and the `unwrap` depending on it are colocated. That also removes the potential it gets called from somewhere the invariant doesn't hold. WDYT?

Reply:
I kinda like this small function, since it's called in two places which should always do the precise same thing. I've addressed this by eliminating the unwrap by making it return an `Option`, and removing the hardcoded `Some`s from the two call sites